src/libstd/sys/windows/os.rs

```diff
@@ -32,17 +32,40 @@ pub fn errno() -> i32 {
 }
 
 /// Gets a detailed string description for the given error number.
-pub fn error_string(errnum: i32) -> String {
+pub fn error_string(mut errnum: i32) -> String {
     // This value is calculated from the macro
     // MAKELANGID(LANG_SYSTEM_DEFAULT, SUBLANG_SYS_DEFAULT)
     let langId = 0x0800 as c::DWORD;
 
+    // NTSTATUS errors may be encoded as HRESULT, see [MS-ERREF].
+    // [MS-ERREF]: https://msdn.microsoft.com/en-us/library/cc231198.aspx
+    const NTSTATUS_BIT: i32 = 0x1000_0000;
+
     let mut buf = [0 as c::WCHAR; 2048];
 
     unsafe {
-        let res = c::FormatMessageW(c::FORMAT_MESSAGE_FROM_SYSTEM |
+        let (module, flags) = if (errnum & NTSTATUS_BIT) != 0 {
+            // format according to https://support.microsoft.com/en-us/help/259693
+            const NTDLL_DLL: &'static [u16] = &['N' as _, 'T' as _, 'D' as _, 'L' as _, 'L' as _,
+                                                '.' as _, 'D' as _, 'L' as _, 'L' as _, 0];
+            let module = c::GetModuleHandleW(NTDLL_DLL.as_ptr());
+
+            if module == ptr::null_mut() {
+                // Ok to call recursively, GetModuleHandleW doesn't return NTSTATUS
+                return format!("OS Error {} (GetModuleHandleW() returned error {})",
```

Comment:
Sure yeah, but that's an error about how it can't be formatted. This is an error about how an intermediate step failed, so this message could be confusing. I think if the ptr coming out here is null we should just fall through to the same code path as below